File: validate.py
import json, os, sys
from jsonschema.validators import Draft202012Validator as V

BASE = os.path.dirname(os.path.abspath(__file__))

def load(p): return json.load(open(os.path.join(BASE, p)))

core = load("eventcal.schema.json")
registry = load("registry.json")["types"]
instances = load("instances.json")

V.check_schema(core)
core_validator = V(core, format_checker=V.FORMAT_CHECKER)

# Cache de validadores de extension por type
ext_cache = {}
def ext_validator(t):
    if t not in ext_cache:
        spec = registry[t]
        sch = load(spec["file"])
        V.check_schema(sch)
        ext_cache[t] = V(sch, format_checker=V.FORMAT_CHECKER)
    return ext_cache[t]

def validate_doc(doc):
    # Fase 1: esqueleto contra el nucleo
    errs = sorted(core_validator.iter_errors(doc), key=lambda e: e.path)
    if errs:
        e = errs[0]
        loc = "/".join(str(p) for p in e.absolute_path) or "(raiz)"
        return ("fase 1 (nucleo)", f"{loc}: {e.message[:70]}")
    # Fase 2: details de cada evento contra su extension
    for i, ev in enumerate(doc.get("events", [])):
        t = ev.get("type")
        if not t:
            continue
        if t not in registry:
            # Tipo desconocido: pasa el esqueleto, sin validacion profunda
            continue
        details = ev.get("details", {})
        derrs = sorted(ext_validator(t).iter_errors(details), key=lambda e: e.path)
        if derrs:
            e = derrs[0]
            loc = f"events/{i}/details/" + "/".join(str(p) for p in e.absolute_path)
            return (f"fase 2 ({t})", f"{loc.rstrip('/')}: {e.message[:60]}")
    return (None, None)

print("Validacion en dos fases (nucleo + registro)\n")
print(f"{'instancia':<28} {'esperado':<9} {'resultado':<11} {'donde fallo'}")
print("-"*82)
for name, doc in instances.items():
    phase, msg = validate_doc(doc)
    ok = phase is None
    expected_ok = name.startswith("valid")
    verdict = "OK" if ok == expected_ok else "!! INESPERADO"
    res = "valido" if ok else "rechazado"
    where = "" if ok else phase
    note = " [tipo desconocido: solo esqueleto]" if name == "valid_unknown_type" else ""
    print(f"{name:<28} {'valido' if expected_ok else 'rechazo':<9} {res:<11} {where}{('  '+verdict) if verdict!='OK' else ''}{note}")
    if not ok and not expected_ok:
        print(f"      -> {msg}")

File: validate.rb
# frozen_string_literal: true
#
# Validador de dos fases para eventcal/v1 usando json_schemer (draft 2020-12).
# Gemfile:  gem 'json_schemer'
#
# Fase 1: el documento completo contra el nucleo (esqueleto).
# Fase 2: el bloque `details` de cada evento contra el esquema de su `type`,
#         resuelto via registry.json. Tipo no registrado => solo esqueleto.
#
# Nota: en produccion esto vive en un service object de Rails
# (p.ej. app/services/eventcal/validator.rb), no como script suelto.

require 'json'
require 'json_schemer'
require 'pathname'

module Eventcal
  class Validator
    BASE = Pathname.new(__dir__)

    def initialize
      # format: true para que uri/date-time/date se asserten de verdad
      @core = JSONSchemer.schema(BASE.join('eventcal.schema.json'), format: true)
      @registry = JSON.parse(BASE.join('registry.json').read)['types']
      @ext = {} # cache de validadores de extension por type
    end

    # Devuelve [] si es valido, o [fase, mensaje] con el primer error.
    def validate(doc)
      # --- Fase 1: esqueleto contra el nucleo ---
      if (e = first_error(@core, doc))
        return ['fase 1 (nucleo)', e]
      end

      # --- Fase 2: details por tipo ---
      Array(doc['events']).each_with_index do |evt, i|
        type = evt['type']
        next if type.nil?
        next unless @registry.key?(type) # tipo desconocido: pasa solo el esqueleto

        if (e = first_error(extension_for(type), evt['details'] || {}))
          return ["fase 2 (#{type})", "events/#{i}/details#{e}"]
        end
      end

      [] # valido
    end

    def valid?(doc)
      validate(doc).empty?
    end

    private

    def extension_for(type)
      @ext[type] ||= JSONSchemer.schema(
        BASE.join(@registry[type]['file']), format: true
      )
    end

    # Primer error en formato "pointer: mensaje", o nil si valida.
    def first_error(schemer, data)
      err = schemer.validate(data).first
      err && "#{err['data_pointer']}: #{err['error']}"
    end
  end
end

# Uso de ejemplo:
# doc = JSON.parse(File.read('mi_calendario.json'))
# phase, msg = Eventcal::Validator.new.validate(doc)
# puts phase ? "rechazado en #{phase}: #{msg}" : 'valido'
